File: 3306_count_substrings.py
def solve(word, k):
    def get_prefix_sum(word, alpha):
        dp = [0]
        for al in word:
            if al == alpha:
                dp.append(dp[-1] + 1)
            else:
                dp.append(dp[-1])
        return dp

    def is_valid(left, right, store):
        right += 1
        count = 0

        for alpha in "aeiou":
            alpha_count = store[alpha][right] - store[alpha][left]
            if alpha_count < 1:
                return -1
            count += alpha_count

        length = right - left
        cons_count = length - count
        print("\t\t", word[left:right], cons_count)
        return cons_count

    def get_max_left(r):
        print(f"\t from get_max_left:")
        ans = None
        left, right = 0, r
        while left <= right:
            mid = left + (right - left) // 2
            cons_count = is_valid(mid, r, store)
            if cons_count < k:
                right = mid - 1
            elif cons_count == k:
                if ans is None:
                    ans = mid
                ans = min(ans, mid)
                right = mid - 1
            else:
                left = mid + 1
        return ans

    def get_min_left(r):
        print(f"\t from get_min_left:")
        ans = None
        left, right = 0, r
        while left <= right:
            mid = left + (right - left) // 2
            cons_count = is_valid(mid, r, store)
            if cons_count < k:
                right = mid - 1
            elif cons_count == k:
                if ans is None:
                    ans = mid
                ans = max(ans, mid)
                left = mid + 1
            else:
                left = mid + 1
        return ans

    store = dict((alpha, get_prefix_sum(word, alpha)) for alpha in "aeiou")
    ans = 0
    for right in range(k-1, len(word)):
        print(f"For right = {right}:")
        min_left = get_max_left(right)
        max_left = get_min_left(right)

        if max_left is None or min_left is None: continue
        ans += max_left - min_left + 1
        print(f"\t min_left = {min_left}, max_left = {max_left}")

        print()
        print()
        print()
    return ans


print(solve("iqeaouqi", 2))

File: 3306_count_substrings.cpp
#include <iostream>
#include <map>

using namespace std;

long long helper(string word, int k) {
  long long count = 0;

  map<char, int> vowel_count = {
      {'a', 0}, {'e', 0}, {'i', 0}, {'o', 0}, {'u', 0}};
  int consonant_count = 0;

  int left = 0, right = 0;
  while (right < word.size()) {
    if (vowel_count.count(word[right]))
      vowel_count[word[right]]++;
    else
      consonant_count++;

    bool flg = true;
    for (auto item : vowel_count)
      flg &= item.second > 0;

    while (flg && consonant_count >= k) {
      count += word.size() - right;
      if (vowel_count.count(word[left]))
        vowel_count[word[left]]--;
      else
        consonant_count--;

      for (auto item : vowel_count)
        flg &= item.second > 0;

      left++;
    }
    right++;
  }

  return count;
}

long long solve(string word, int k) {
  return helper(word, k) - helper(word, k + 1);
}

int main() {
  cout << solve("aeioqq", 1) << endl;
  cout << solve("aeiou", 0) << endl;
  cout << solve("ieaouqqieaouqq", 1) << endl;
  cout << solve("iqeaouqi", 2) << endl;
}
